File: sdk/go/mics-hook-sdk/micshooksdk/signing.go
package micshooksdk

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/binary"
)

func ComputeHookSignBase64(tenantSecret string, payloadBytes []byte, requestID string, timestampMs int64) string {
	var tsBytes [8]byte
	binary.LittleEndian.PutUint64(tsBytes[:], uint64(timestampMs))

	mac := hmac.New(sha256.New, []byte(tenantSecret))
	mac.Write(payloadBytes)
	mac.Write([]byte(requestID))
	mac.Write(tsBytes[:])
	return base64.StdEncoding.EncodeToString(mac.Sum(nil))
}

func VerifyHookSignBase64(tenantSecret string, payloadBytes []byte, requestID string, timestampMs int64, signBase64 string) bool {
	if signBase64 == "" {
		return false
	}
	if !isCanonicalBase64(signBase64) {
		return false
	}

	expected := ComputeHookSignBase64(tenantSecret, payloadBytes, requestID, timestampMs)
	return constantTimeEqualBase64(expected, signBase64)
}

func ComputeMqEventSignBase64(tenantSecret string, payloadBytes []byte) string {
	mac := hmac.New(sha256.New, []byte(tenantSecret))
	mac.Write(payloadBytes)
	return base64.StdEncoding.EncodeToString(mac.Sum(nil))
}

func VerifyMqEventSignBase64(tenantSecret string, payloadBytes []byte, signBase64 string) bool {
	if signBase64 == "" {
		return false
	}
	if !isCanonicalBase64(signBase64) {
		return false
	}
	expected := ComputeMqEventSignBase64(tenantSecret, payloadBytes)
	return constantTimeEqualBase64(expected, signBase64)
}

func constantTimeEqualBase64(a, b string) bool {
	ab, err := base64.StdEncoding.DecodeString(a)
	if err != nil {
		return false
	}
	bb, err := base64.StdEncoding.DecodeString(b)
	if err != nil {
		return false
	}
	if len(ab) != len(bb) {
		return false
	}
	return hmac.Equal(ab, bb)
}

func isCanonicalBase64(s string) bool {
	decoded, err := base64.StdEncoding.DecodeString(s)
	if err != nil {
		return false
	}
	return base64.StdEncoding.EncodeToString(decoded) == s
}

File: sdk/node/mics-hook-sdk/src/signing.ts
import crypto from "node:crypto";
import { MqEventCodec, type MqEvent } from "./proto/mics_hook.js";

function toLeI64Bytes(value: number): Uint8Array {
  const buf = Buffer.alloc(8);
  buf.writeBigInt64LE(BigInt(value), 0);
  return buf;
}

export function computeHookSignBase64(tenantSecret: string, meta: { requestId: string; timestampMs: number }, payloadBytes: Uint8Array): string {
  const key = Buffer.from(tenantSecret ?? "", "utf8");
  const requestIdBytes = Buffer.from(meta.requestId ?? "", "utf8");
  const tsBytes = toLeI64Bytes(meta.timestampMs ?? 0);

  return crypto.createHmac("sha256", key).update(payloadBytes).update(requestIdBytes).update(tsBytes).digest("base64");
}

export function verifyHookSign(
  tenantSecret: string,
  meta: { requestId: string; timestampMs: number },
  payloadBytes: Uint8Array,
  signBase64: string
): boolean {
  if (!signBase64) return false;
  const expected = computeHookSignBase64(tenantSecret, meta, payloadBytes);
  return constantTimeEqualBase64(expected, signBase64);
}

function computeMqBytesSignBase64(tenantSecret: string, payloadBytes: Uint8Array): string {
  const key = Buffer.from(tenantSecret ?? "", "utf8");
  return crypto.createHmac("sha256", key).update(payloadBytes).digest("base64");
}

export function computeMqEventSignBase64(tenantSecret: string, event: MqEvent): string {
  const bytes = MqEventCodec.encode({ ...event, sign: "" }).finish();
  return computeMqBytesSignBase64(tenantSecret, bytes);
}

export function verifyMqEventSign(tenantSecret: string, event: MqEvent, signBase64: string): boolean {
  if (!signBase64) return false;
  const expected = computeMqEventSignBase64(tenantSecret, event);
  return constantTimeEqualBase64(expected, signBase64);
}

function constantTimeEqualBase64(a: string, b: string): boolean {
  try {
    if (!isCanonicalBase64(b)) return false;
    const ab = Buffer.from(a, "base64");
    const bb = Buffer.from(b, "base64");
    if (ab.length !== bb.length) return false;
    return crypto.timingSafeEqual(ab, bb);
  } catch {
    return false;
  }
}

function isCanonicalBase64(s: string): boolean {
  const buf = Buffer.from(s, "base64");
  return buf.toString("base64") === s;
}
